multicore_runtime/wsgi.py

```diff
@@ -0,0 +1,128 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""Configure a user project and instantiate WSGI app meta_app to serve it.
+
+Importing this module will result in side-effects, such as registering the
+project's default ticket via vmstub.Register. This is a broadly compatible way
+to induce the webserver to run initialization code before starting to serve the
+WSGI app.
+"""
+
+import logging
+import os
+
+import cloud_logging
+from dispatcher import dispatcher
+from middleware import health_check_middleware
+from middleware import reset_environment_middleware
+from wsgi_config import env_vars_from_env_config
+from wsgi_config import get_module_config
+from wsgi_config import get_module_config_filename
+from wsgi_config import load_user_scripts_into_handlers
+from wsgi_config import ThreadLocalDict
+from wsgi_config import user_env_vars_from_appinfo
+
+from google.appengine.ext.vmruntime import vmconfig
+from google.appengine.ext.vmruntime import vmstub
+
+# Configure logging to output structured JSON to Cloud Logging.
+root_logger = logging.getLogger('')
+try:
+  handler = cloud_logging.CloudLoggingHandler()
+  root_logger.addHandler(handler)
+except IOError:
+  # If the Cloud Logging endpoint does not exist, just use the default handler
+  # instead. This will be the case when running in local dev mode.
+  pass
+
+root_logger.setLevel(logging.INFO)
+
+# Fetch application configuration via the config file.
+appinfo = get_module_config(get_module_config_filename())
+env_config = vmconfig.BuildVmAppengineEnvConfig()
+
+# Ensure API requests include a valid ticket by default.
+vmstub.Register(vmstub.VMStub(env_config.default_ticket))
+
+# Take an immutable snapshot of env data from env_config. This is added to the
+# environment in `reset_environment_middleware` in a particular order to ensure
+# that it cannot be overridden by other steps.
+frozen_env_config_env = tuple(
+    env_vars_from_env_config(env_config).iteritems())
+
+# Also freeze user env vars specified in app.yaml. Later steps to modify the
+# environment such as env_vars_from_env_config and request middleware
+# will overwrite these changes. This is added to the environment in
+# `reset_environment_middleware`.
+frozen_user_env = tuple(
+    user_env_vars_from_appinfo(appinfo).iteritems())
+
+# While the primary use of the frozen env vars is for
+# `reset_environment_middleware`, we'll also add them to the env here to make
+# them available during app preloading. This is useful for apps which have
+# side-effects at import time (which is not recommended).
+os.environ.update(frozen_user_env)
+os.environ.update(frozen_env_config_env)
+
+# Decide whether to enable support for legacy end-to-end tests. This is intended
+# to be temporary.
+# TODO(apphosting): Modify the end-to-end test runner to make this step
+# unnecessary.
+legacy_e2e = appinfo.vm_settings.get('vm_runtime') == 'python'
+
+# Load user code.
+if legacy_e2e:
```

Comment:
Only because I don't feel that `appinfo.vm_settings.get('vm_runtime') == 'python'` is a good API for this, and my instinct is to keep bad code and bad APIs isolated from other code as much as possible until it can be fixed. But in this case the whole blob has the same problem, so I should just integrate them.